pkg/metrics/sketch_series.go

```diff
@@ -78,6 +79,117 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
+// The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
+func (sl SketchSeriesList) MarshalSplitCompress() ([]*[]byte, error) {
+	input := bytes.NewBuffer(make([]byte, 0, 1024))
+	output := bytes.NewBuffer(make([]byte, 0, 1024))
+
+	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
+	footer := []byte{0x12, 0}
+
+	compressor, e := stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+	if e != nil {
+		return nil, e
+	}
+	payloads := []*[]byte{}
+	precompressionBuf := make([]byte, 1024)
+
+	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
+	for _, ss := range sl {
+		if len(ss.Points) > cap(dsl) {
+			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+		}
+
+		for i, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl[i] = gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			}
+		}
+
+		sketch := gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl[:len(ss.Points)],
+		}
+
+		// Pack the protobuf metadata
+		i := 0
+		precompressionBuf[i] = 0xa
+		i++
+		i = encodeVarintAgentPayload(precompressionBuf, i, uint64(sketch.Size()))
+
+		// Resize the pre-compression buffer if needed
+		totalItemSize := sketch.Size() + i
+		if totalItemSize > cap(precompressionBuf) {
+			precompressionBuf = append(precompressionBuf, make([]byte, totalItemSize-cap(precompressionBuf))...)
+			precompressionBuf = precompressionBuf[:cap(precompressionBuf)]
+		}
+
+		// Marshal the sketch to the precompression buffer
+		_, e := sketch.MarshalTo(precompressionBuf[i:])
+		if e != nil {
+			return nil, e
+		}
+
+		// Compress the protobuf metadata and the marshaled sketch
+		switch compressor.AddItem(precompressionBuf[:totalItemSize]) {
+		case stream.ErrItemTooBig, stream.ErrPayloadFull:
+			// Since the compression buffer is full - flush it and rotate
+			payload, e := compressor.Close()
+			if e != nil {
+				return nil, e
+			}
+			payloads = append(payloads, &payload)
+			input.Reset()
+			output.Reset()
+			compressor, e = stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+			if e != nil {
+				return nil, e
+			}
+
+			// Add it to the new compression buffer - since this is a new compressor it should never overflow.
+			e = compressor.AddItem(precompressionBuf[:totalItemSize])
+			if e != nil {
+				return nil, e
+			}
+		}
+	}
+
+	payload, e := compressor.Close()
+	if e != nil {
+		return nil, e
+	}
+	payloads = append(payloads, &payload)
+
+	// return payloads, nonCompressed
+	return payloads, nil
+}
+
+// taken from agent_payload.pb.go
+func encodeVarintAgentPayload(dAtA []byte, offset int, v uint64) int {
```

Comment:
Will this be compatible with future protobuf version ? if not we should add a test for this.

Reply:
The UTs I added cover comparing `SketchSeriesList.Marshal()` to the new logic I [added here](https://github.com/DataDog/datadog-agent/pull/7312/files#diff-20e25bc6d4eb3cdad003697152280f8672e99f1bac21c0af9b253359e4189669R126)  - so if the protobuf binary format ever changes, this test will fail.